lib/logdna/client.rb

```diff
@@ -1,164 +1,130 @@
 
-      check_side_buffer
-      buffer_size = @buffer.write(msg)
-      @messages.push(message_hash(msg, opts))
-      buffer_size
-    end
+        @lock.unlock if @lock.locked?
 
-    def queue_to_buffer(queue=@queue)
-      next_object = queue.shift
-      write_to_buffer(next_object[:msg], next_object[:opts])
-    end
+        flush if @flush_limit <= @buffer_byte_size
 
-    def process_buffer(buffer_size)
-      if buffer_size > @actual_byte_limit
-        @buffer_over_limit = true
-        flush()
-        @buffer_over_limit = false
+        schedule_flush unless @flush_scheduled
       else
-        @task = create_flush_task
+        @side_messages.push(process_message(msg, opts))
       end
     end
 
-    # this should be running synchronously if @buffer_over_limit i.e. called from self.buffer
-    # else asynchronously through @task
-    def flush()
-      if defined? @@request and !@@request.nil?
-        request_messages = []
-        @lock.synchronize do
-          request_messages = @messages
-          @buffer.truncate(0)
-          @messages = []
-        end
-        return if request_messages.empty?
+    def send_request
+      @request.body = {
+        e: "ls",
+        ls: @buffer.concat(@side_messages)
+      }.to_json
+      @side_messages.clear
 
-        real = {
-          e: 'ls',
-          ls: request_messages,
-        }.to_json
-
-        @@request.body = real
-        @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
-          http.request(@@request)
+      begin
+        @response = Net::HTTP.start(
+          @uri.hostname,
+          @uri.port,
+          use_ssl: @uri.scheme == "https"
+        ) do |http|
+          http.request(@request)
         end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+        if @response.is_a?(Net::HTTPForbidden)
+          puts "Please provide a valid ingestion key"
+        elsif !@response.is_a?(Net::HTTPSuccess)
+          puts "The response is not successful "
         end
+
+        @exception_flag = false
+      rescue SocketError
+        puts "Network connectivity issue"
+        @exception_flag = true
+        @side_messages.concat(@buffer)
+      rescue Errno::ECONNREFUSED => e
+        puts "The server is down. #{e.message}"
+        @exception_flag = true
+        @side_messages.concat(@buffer)
+      rescue Timeout::Error => e
+        puts "Timeout error occurred. #{e.message}"
+        @exception_flag = true
+        @side_messages.concat(@buffer)
+      ensure
+        @buffer.clear
+        @lock.unlock if @lock.locked?
       end
     end
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
-        flush()
+    def flush
+      @flush_scheduled = false
+      return if @buffer.empty?
```

Comment:
You got rid of the side buffer check - what happens if there are messages in `@side_messages`?